Unix: Default to a users-specific temp directory for extracting single-file apps

swaroop-sridhar opened the pull request:
In .net core 3, single file apps run by extracting the bundled contents to a temp directory.
The extraction directory is machine specific, and can be set through DOTNET_BUNDLE_EXTRACT_BASE_DIR environment variable.

When this setting is not configured, the host tries to use certain default directories.
On windows, extraction is within %TMPDIR%, which is user specific.
On Unix systems $TMPDIR/.net if set, which may be user specific (ex: MAC)
Otherwise, the extraction directory is within /var/tmp/ or /tmp/ which is common to all users, and may be locked by a specific user on first creation.

Therefore, this change fixes this issue by defaulting the extraction base directory in Unix systems to
`<temp-dir>/.net/<user-ID>` , where
`<temp-dir>/.net/` has permission 0777, and
`<temp-dir>/.net/<user-ID>/` has permission 0700.

This fix will be migrated to coreclr/3.1 branch for servicing.

Testing: Manual testing on Unix/Mac systems, since we don't have the setup to add automated tests with multiple users.

Fixes https://github.com/dotnet/core-setup/issues/8882

swaroop-sridhar committed:
Unix: Default to a users-specific temp directory for extracting single-file apps

In .net core 3, single file apps run by extracting the bundled contents to a temp directory.
The extraction directory is machine specific, and can be set through DOTNET_BUNDLE_EXTRACT_BASE_DIR environment variable.

When this setting is not configured, the host tries to use certain default directories.
On windows, extraction is within %TMPDIR%, which is user specific.
On Unix systems $TMPDIR/.net if set, which may be user specific (ex: MAC)
Otherwise, the extraction directory is within /var/tmp/ or /tmp/ which is common to all users, and may be locked by a specific user on first creation.

Therefore, this change fixes this issue by defaulting the extraction base directory in Unix systems to
`<temp-dir>/.net/<user-ID>` , where
`<temp-dir>/.net/` has permission 0777, and
`<temp-dir>/.net/<user-ID>/` has permission 01700.

This fix will be migrated to coreclr/3.1 branch for servicing.

Testing: Manual testing on Unix/Mac systems, since we don't have the setup to add automated tests with multiple users.

Fixes https://github.com/dotnet/core-setup/issues/8882

jkotas commented:
https://github.com/dotnet/designs/pull/88

lpereira commented on src/installer/corehost/common/pal.unix.cpp:
```diff
@@ -331,6 +331,38 @@ bool pal::get_temp_directory(pal::string_t& tmp_dir)
+    }
+
+    // Create $TMPDIR/.net/username accessible only to the user
+    if (::mkdir(extraction_dir, 0700) != 0 && errno != EEXIST)
```
I'd also set S_ISVTX (`01000` on Linux, but might be different on other systems) in these two `mkdir()` calls in addition to their access permissions.  This is the sticky bit.

With this bit set, only the owner of `extraction_dir` can rename or delete that directory.  (Otherwise, there's a potential race condition here: while you can only remove empty directories -- and need to recursively delete files from a directory -- a malicious process can remove the directory even before we were able to start extracting the files there.)

swaroop-sridhar replied:
Thanks @lpereira, I've fixed this on the `extraction_dir`, per our discussion.

swaroop-sridhar commented:
@lpereira  I've addressed your comments about `S_ISVTX` bit, and made another change to replace the `getlogin()` call, in the last two commits. Please take a look. Thanks,.

lpereira commented on src/installer/corehost/common/pal.unix.cpp:
```diff
@@ -331,6 +332,55 @@ bool pal::get_temp_directory(pal::string_t& tmp_dir)
+    pal::string_t dotnetdir(extraction_dir);
+
+    // getuid() is the real user ID, and the call has no defined errors.
+    struct passwd* passwd = getpwuid(getuid());
```
Is there a change `getpwuid()` can be called by another thread?

Also, nit:
```suggestion
    const passwd* passwd = getpwuid(getuid());
```

(No need to say `struct` in C++, it's always implied in type declarations.)

swaroop-sridhar replied:
> Is there a change `getpwuid()` can be called by another thread?

`getpwuid()` call here is within the `get_default_bundle_extraction_base_dir()` specifically run only at apphost start. So, there's no other thread running. 

I thought of using `getpwuid_r(...)` but realized we'll have bigger problems if something else starts running before the contents are extracted.

lpereira commented on src/installer/corehost/common/pal.unix.cpp:
```diff
@@ -331,6 +332,55 @@ bool pal::get_temp_directory(pal::string_t& tmp_dir)
+    else if (errno != EEXIST)
+    {
+        return false;
+    }
```
What happens if you run a program with `$TMPDIR/.net` already created, but with the permissions different than the one you set with `chmod()` because someone went there and changed them?

swaroop-sridhar replied:
That would be too bad. We just provide the `$TPMDIR/.net/$UID` as the default for `$DOTNET_BUNDLE_EXTRACTION_BASE_DIR`. If the default directory is not available/accessible on a system, then the user should set `DOTNET_BUNDLE_EXTRACTION_BASE_DIR` to an alternate location.

omajid commented:
Can you help me understand why we are using an error prone approach where `/tmp/.net` (with a  predictable name) is used? Wouldn't it be more secure to fall back to a user-specific path? `$XDG_CACHE_HOME` (falling back to `$HOME/.cache/`) seems like a location that would be much less susceptible attacks and accidental clashes. 

Also, most recent distributions treat `/tmp` as temporary. `systemd`, for example, wipes things older than 10 days on `/tmp/` on my machine. Compare this with the persistent `$HOME/.cache`.

danmosemsft commented:
@omajid might be best to open a new issue so folks see it.
